libexec/rtld-elf/aarch64/rtld_machdep.h

```diff
@@ -36,23 +37,145 @@
+#endif
+
 struct Struct_Obj_Entry;
 
 /* Return the address of the .dynamic section in the dynamic linker. */
+#ifdef __CHERI_PURE_CAPABILITY__
+#define rtld_dynamic(obj) (&_DYNAMIC)
+#else
 #define	rtld_dynamic(obj)						\
 ({									\
 	Elf_Addr _dynamic_addr;						\
 	asm volatile("adr	%0, _DYNAMIC" : "=&r"(_dynamic_addr));	\
 	(const Elf_Dyn *)_dynamic_addr;					\
 })
+#endif
 
 Elf_Addr reloc_jmpslot(Elf_Addr *where, Elf_Addr target,
     const struct Struct_Obj_Entry *defobj, const struct Struct_Obj_Entry *obj,
     const Elf_Rel *rel);
 
+#if __has_feature(capabilities)
+
+#define	FUNC_PTR_REMOVE_PERMS						\
+	(CHERI_PERM_SEAL | CHERI_PERM_STORE | CHERI_PERM_STORE_CAP |	\
+	CHERI_PERM_STORE_LOCAL_CAP)
+
+#define	DATA_PTR_REMOVE_PERMS						\
+	(CHERI_PERM_SEAL | CHERI_PERM_EXECUTE)
+
+/* TODO: we should have a separate member for .text/rodata */
+#define get_codesegment(obj) ((obj)->text_rodata_cap)
+
+/* TODO: ABIs with tight bounds */
+#define can_use_tight_pcc_bounds(obj) ((void)(obj), false)
+
+#define set_bounds_if_nonnull(ptr, size)	\
+	do { if (ptr) { ptr = cheri_setbounds_sametype(ptr, size); } } while(0)
+
+#endif /* __has_feature(capabilities) */
+
+#if __has_feature(capabilities)
+
+typedef void (* __capability dlfunccap_t)(struct __dlfunc_arg);
+
+/*
+ * Create a pointer to a function.
+ */
+static inline dlfunccap_t
+make_code_cap_pointer(const Elf_Sym *def, const struct Struct_Obj_Entry *defobj,
+    bool tight_bounds, size_t addend)
+{
+#ifdef __CHERI_PURE_CAPABILITY__
+	const void *ret = get_codesegment(defobj) + def->st_value;
+#else
+	const void * __capability ret = cheri_setoffset(cheri_getdefault(),
+	    (uintptr_t)(defobj->relocbase + def->st_value));
+#endif
+	/* Remove store and seal permissions */
+	ret = cheri_clearperm(ret, FUNC_PTR_REMOVE_PERMS);
+	if (tight_bounds) {
+		ret = cheri_setbounds(ret, def->st_size);
+	}
+	/*
+	 * Note: The addend is required for C++ exceptions since capabilities
+	 * for catch blocks point to the middle of a function.
+	 */
+	ret = cheri_incoffset(ret, addend);
+	/* All code pointers should be sentries: */
+	ret = __builtin_cheri_seal_entry(ret);
+	return __DECONST_CAP(dlfunccap_t, ret);
+}
+
+static inline dlfunccap_t
+make_function_cap_pointer_with_addend(
+    const Elf_Sym *def, const struct Struct_Obj_Entry *defobj, size_t addend)
+{
+	/* TODO: ABIs with tight bounds */
+	return make_code_cap_pointer(def, defobj, /*tight_bounds=*/false, addend);
+}
+
+static inline dlfunccap_t
+make_function_cap_pointer(const Elf_Sym *def, const struct Struct_Obj_Entry *defobj)
+{
+	return make_function_cap_pointer_with_addend(def, defobj, /*addend=*/0);
+}
+
+static inline void*__capability
+make_data_cap_pointer(const Elf_Sym *def, const struct Struct_Obj_Entry *defobj)
+{
+#ifdef __CHERI_PURE_CAPABILITY__
+	void *ret = defobj->relocbase + def->st_value;
+#else
+	void * __capability ret = cheri_setoffset(cheri_getdefault(),
+	    (uintptr_t)(defobj->relocbase + def->st_value));
+#endif
```

Comment:
Could either have a get_datasegment or inline it as:
```suggestion
	void * __capability relocbase;
#ifdef __CHERI_PURE_CAPABILITY__
	relocbase = defobj->relocbase;
#else
	relocbase = cheri_getdefault() + (vaddr_t)defobj->relocbase;
#endif
	void * __capability ret = relocbase + def->st_value;
```